### PR TITLE
Z layers

### PR DESCRIPTION
**Pull request by koosemose**
Implemented multiple layers.
Movement between layers is currently hardcoded to Home and End (until we have a more featured keyboard controller)
There is a bit of parallax between layers, along with darker shading the farther below you a layer is.
A few caveats, building on lower levels is currently only possible via dev mode (character's don't pathfind there yet), and you can only place a single tile/furniture at a time (due to interaction with the pathfinding system.
I've been advised to leave pathfinding as is, since there are other's already at work on the system, so the less than perfect interactions with Pathfinding and graph building will have to stay for now.

### CODE REVIEW

**Comment by vogonistic** [comment]
You already have conflicts =)

**Comment by koosemose** [comment]
Yep, that's what happens when you try to fix conflicts just as a Gatekeeper happens to be merging.

**Comment by Sommerbo** [comment]
Tested this PR. Nothing broke, and this looks fantastic!

**Comment by Ohda** [comment]
Tested without detecting a bug.

File: Assets/Scripts/Controllers/CameraController.cs [hunk 1]
```diff
@@ -29,6 +46,7 @@ public void Update(bool modal)
         {
```
Comment (Grenkin1988) [inline Assets/Scripts/Controllers/CameraController.cs]: `Camera.main.orthographicSize.AreEqual(zoomTarget) == false`

File: Assets/Scripts/Controllers/CameraController.cs [hunk 2]
```diff
@@ -44,4 +62,83 @@ public void ChangeZoom(float amount)
...
+            for (int i = 0; i < layerCameras.Length; i++)
+            {
+                if (i < newLayer)
+                {
+                    layerCameras[i].gameObject.SetActive(false);
+                }
+                else
+                {
+                    layerCameras[i].gameObject.SetActive(true);
+                }
```
Comment (Grenkin1988) [inline Assets/Scripts/Controllers/CameraController.cs]: this
```
                if (i < newLayer)
                {
                    layerCameras[i].gameObject.SetActive(false);
                }
                else
                {
                    layerCameras[i].gameObject.SetActive(true);
                }
```
could be written as this
`layerCameras[i].gameObject.SetActive(i >= newLayer);`

File: Assets/Scripts/Controllers/CameraController.cs [hunk 3]
```diff
@@ -44,4 +62,83 @@ public void ChangeZoom(float amount)
...
+
+        // We don't have the right number of cameras for our layers
+        if (layerCameras == null || layerCameras.Length != WorldController.Instance.World.Depth)
+        {
+            int depth = WorldController.Instance.World.Depth;
+            layerCameras = new Camera[depth];
+            for (int i = 0; i < depth; i++)
+            {
+                // This creates a new GameObject and adds it to our scene.
+                GameObject camera_go = new GameObject();
```
Comment (Grenkin1988) [inline Assets/Scripts/Controllers/CameraController.cs]: -> `cameraGameObject`

File: Assets/Scripts/Controllers/TileSpriteController.cs [hunk 4]
```diff
@@ -74,6 +77,14 @@ protected override void OnChanged(Tile tile)
         }
         
         tile_go.GetComponent<SpriteRenderer>().sprite = SpriteManager.current.GetSprite("Tile", tile.Type.Name);
+        if (tile.Type == TileType.Empty)
```
Comment (Grenkin1988) [inline Assets/Scripts/Controllers/TileSpriteController.cs]: `tile_go.SetActive(tile.Type != TileType.Empty);`

File: Assets/Scripts/Models/World.cs [hunk 5]
```diff
@@ -216,13 +219,16 @@ public void RandomizeTiles()
         {
             for (int y = 0; y < Height; y++)
             {
-                if (UnityEngine.Random.Range(0, 2) == 0)
-                {
-                    tiles[x, y].Type = TileType.Empty;
-                }
-                else
+                for (int z = 0; z < Depth; z++)
```
Comment (Grenkin1988) [inline Assets/Scripts/Models/World.cs]: It seems that this method ( `RandomizeTiles()` ) is not used anywhere.
Should we delete it?

**Comment by Grenkin1988** [comment]
Looks very nice. Good job. :+1: 
Have just minor cosmetic suggestions. ^

File: Assets/Scripts/Models/World.cs (hunk 5 above)
Comment (koosemose) [inline Assets/Scripts/Models/World.cs]: Not sure, this is a pretty massive change to the game, that touches on some major classes, I didn't want to mess with more than absolutely necessary to get this functional.

**Comment by vogonistic** [comment]
Merged as is! I've been looking forward to this!

**Pull request merged**